[WIP] Code style: preview of Uncrustify formatting

ArseniyShestakov opened the pull request:
This is preview only! 

There some things I personally still want to talk about:

- Switching to no space after * and &.
- Namespace indention. Some parts of our code look especially ugly when it's disabled.

Roadblocks:

- [ ] Problem with formatting of ```**``` and ```***``` @FeniksFire pointed to:
```
tiles = new CTileInfo * *[width];
int main(int argc, char * * argv)
```
- [x] Problem with detection of lambdas with omitted parameters list. Should be fixed by #345.
- [ ] Indention of lambda expressions. I hack-fixed them with "indent_continue = -1", but that not a solution.
- [ ] Multi-line if-else conditions. I'll talk about them in comments.
- [ ] Preprocessor directives formatting. I didn't look a them at all.
- [ ] Changes in formatting multi-line commented-out code. I don't want Uncrustify to touch it.
- [ ] And really multi-line everything since "indent_continue = -1" tend to cause weird results.

What would need to be fixed manually after initial formatting:

- In switch break should be moved inside curly brackets. Can be done with Uncrustify option "mod_move_case_break", but It's better to be enabled only once so it's not break something in future.
- Some tabs will need to be replaced with spaces. Formatter doesn't replace tabs with spaces when it's for instance move colons in class initialization.

Please comment on code changes! Then when it's updated we won't have too many outdated comments.
I'll force push this branch every time with two commits: uncrustify config file and preview.

ArseniyShestakov commented:
Before comment don't forget to check out our coding guidelines:
https://wiki.vcmi.eu/index.php?title=Coding_guidelines
They partially outdated, but we really need to keep them up to date now. 

Feel free to leave any feedback on what you like or dislike about formatted code or if something is missing.

mwu-tow commented on client/gui/CGuiHandler.h:
```diff
@@ -48,33 +49,34 @@ class CGuiHandler
-	               wheelInterested,
-	               doubleClickInterested,
-	               textInterested;
+	rclickable,
```
I'd say that when declaring multiple names, they should be indented, so it is clear that they belong to the type given in the initial declaration line.

ArseniyShestakov replied:
This not really intended. It's problem caused by "indent_continue = -1". But other than that we talked with @alexvins and I come to conclusion it's better if we won't declare using this comma syntax.

Basically it's would be better if they were declared on their own lines.
```
	CIntObjectList lclickable;
	CIntObjectList rclickable;
	CIntObjectList mclickable;
	CIntObjectList hoverable;
	CIntObjectList keyinterested;
	CIntObjectList motioninterested;
	CIntObjectList timeinterested;
	CIntObjectList wheelInterested;
	CIntObjectList doubleClickInterested;
	CIntObjectList textInterested;
```
It's of course will be more code, but it's not as error-prone as declaration with comma especially when it's come to working with pointers.

mwu-tow commented on client/gui/Fonts.cpp:
```diff
@@ -352,27 +355,25 @@ void CBitmapHanFont::renderText(SDL_Surface * surface, const std::string & data,
-    data(CResourceHandler::get()->load(ResourceID("data/" + config["name"].String(), EResType::OTHER))->readAll()),
-    size(config["size"].Float())
+CBitmapHanFont::CBitmapHanFont(const JsonNode & config)
+	: fallback(new CBitmapFont(config["fallback"].String())), data(CResourceHandler::get()->load(ResourceID("data/" + config["name"].String(), EResType::OTHER))->readAll()), size(config["size"].Float())
```
I'm not sure if it is an issue we want to tackle right now, but such long lines are hard to visually parse.
Would it be possible to have a rule like "if initialization list fits within 120 chars then keep it single line, otherwise break it down to single member initializer per line.

Similar sentiment also applies to signatures of function taking many arguments and their call sites.

dydzio0614 replied:
One more reason why it may be good to have one element of initializer list per line.

DjWarmonger replied:
In the end I think it's up to coder to write a line so it's visually clear, using spaces and newlines. There might be simply no "best" set of rules which works for every case imaginable, I think this attempt is just futile.

ArseniyShestakov replied:
@DjWarmonger even if we don't want to enforce it we still need to decide on sane default for source we have now since most old code is badly formatted.

mwu-tow commented on client/windows/CHeroWindow.cpp:
```diff
@@ -103,33 +102,37 @@ CHeroWindow::CHeroWindow(const CGHeroInstance *hero):
 
-	quitButton = new CButton(Point(609, 516), "hsbtns.def", CButton::tooltip(heroscrn[17]), [&]{ close(); }, SDLK_RETURN);
+	quitButton = new CButton(Point(609, 516), "hsbtns.def", CButton::tooltip(heroscrn[17]), [&] { close();
+	}, SDLK_RETURN);
```
Such lambda formatting feels wrong. Either the code block should be fully expanded or everything should be kept inline.

ArseniyShestakov replied:
@mwu-tow this will be fixed once #345 merged. Sorry that I didn't mention it earlier.
Basically Uncrustify have a hard time to parse lambdas with omitted parameter list.

mwu-tow commented on client/windows/CHeroWindow.cpp:
```diff
@@ -300,8 +310,12 @@ void CHeroWindow::update(const CGHeroInstance * hero, bool redrawNeeded)
 {
-	CFunctionList<void()> ony = [=] { close(); };
-	ony += [=] { LOCPLINT->cb->dismissHero(curHero); };
+	CFunctionList<void()> ony = [ = ] {
```
why did default capture specifier get spaces around it? Elsewhere I see `[=]`

ArseniyShestakov replied:
Bug as well and depend on #345.

mwu-tow commented on test/StdInc.h:
```diff
@@ -10,7 +10,5 @@
 	#include <boost/test/unit_test.hpp>
 	#include "CVcmiTestConfig.h"
-	BOOST_GLOBAL_FIXTURE(CVcmiTestConfig);
+BOOST_GLOBAL_FIXTURE(CVcmiTestConfig);
```
I think the intention was to keep whole conditionally compiled block indented.

dydzio0614 commented on AI/StupidAI/StupidAI.cpp:
```diff
@@ -94,20 +96,23 @@ bool isCloser(const EnemyInfo & ei1, const EnemyInfo & ei2, const ReachabilityIn
-	int shooters[2] = {0}; //count of shooters on hexes
+	int shooters[2] =
+	{
+		0
```
Hmm I am not a fan of expanding simple initializers like that

dydzio0614 replied:
@alexvins because if you initialize less elements than exist, rest is zero initialized. This syntax can initialize array of any size with zeros.

dydzio0614 commented on AI/BattleAI/AttackPossibility.cpp:
```diff
@@ -24,21 +24,24 @@ int AttackPossibility::attackValue() const
-	AttackPossibility ap = {enemy, hex, AttackInfo, 0, 0, 0};
+	AttackPossibility ap =
+	{
+		enemy, hex, AttackInfo, 0, 0, 0
```
Also here - not sure if "expand" change like this is actually improving code quality at all

mwu-tow commented on server/NetPacksServer.cpp:
```diff
@@ -22,145 +22,145 @@
-#define ERROR_IF_NOT_OWNS(id)	do{if(!PLAYER_OWNS(id)){WRONG_PLAYER_MSG(gh->getOwner(id)); ERROR_AND_RETURN; }}while(0)
-#define ERROR_IF_NOT(player)	do{if(player != gh->getPlayerAt(c)){WRONG_PLAYER_MSG(player); ERROR_AND_RETURN; }}while(0)
+#define PLAYER_OWNS(id) (gh->getPlayerAt(c) == gh->getOwner(id))
+#define ERROR_AND_RETURN                                                                                                \
```
It would be great if uncrustify could format macro body that:
a) use rules as for formatting code
b) align trailing `\` on the same column

mwu-tow commented on lib/mapping/CMap.h:
```diff
@@ -90,12 +89,9 @@ struct DLL_LINKAGE PlayerInfo
-		h & p7 & hasRandomHero & mainCustomHeroId & canHumanPlay & canComputerPlay & aiTactic & allowedFactions & isFactionRandom &
-				mainCustomHeroPortrait & mainCustomHeroName & heroesNames & hasMainTown & generateHeroAtMainTown &
-				posOfMainTown & team & generateHero;
+		h & p7 & hasRandomHero & mainCustomHeroId & canHumanPlay & canComputerPlay & aiTactic & allowedFactions & isFactionRandom & mainCustomHeroPortrait & mainCustomHeroName & heroesNames & hasMainTown & generateHeroAtMainTown & posOfMainTown & team & generateHero;
```
Another case of "it would be great if overlong lines could be split".

ArseniyShestakov replied:
Thanks for pointing out. I think for the sake of desync debugging we should really put all of serializer statements on own lines:
```
h & p7;
h & hasRandomHero;
h & mainCustomHeroId;
...
```
Probably should submit PR for that.

dydzio0614 commented on AI/StupidAI/StupidAI.cpp:
```diff
@@ -49,7 +49,8 @@ struct EnemyInfo
 	std::vector<BattleHex> attackFrom; //for melee fight
-	EnemyInfo(const CStack * _s) : s(_s), adi(0), adr(0)
+	EnemyInfo(const CStack * _s)
+		: s(_s), adi(0), adr(0)
```
Any advantages of this formatting change except saving some space for really many parameters? For me it's little bit easier to track code when { } belong to something that is written in one line above {, but maybe I freak too much because I saw too many ugly multi-line if's and elses

IvanSavenko replied:
For "really may parameters" one member per line would look much better. Extra bonus - keep commas before member name, e.g.
```
EnemyInfo(const CStack * _s)
: s(_s)
, adi(0)
, adr(0)
```
Or even add vertical alignment (although I doubt that auto-formatters can do this)
```
EnemyInfo (const CStack * _s)
: s       (_s)
, adi     (0)
, adr     (0)
```
Such format helps a lot in readability for classes with 10+ members with names of different length

dydzio0614 replied:
I think we are limited to "binary" solution with uncrustify - either "new line" everything, even for 1,2,3 members or one line 10+ members...

IvanSavenko replied:
New line for everything sounds like a good option for me. And normally you don't need a struct with one member so you always end up with 2+ members

ArseniyShestakov replied:
Yes. We can also don't touch it at all.

@IvanSavenko I'm really against any alignment because it's become complete mess as soon as you use different IDE settings from author of the code. And even if formatter can do some alignment it's will be mess for one developer or another.

dydzio0614 replied:
https://stackoverflow.com/questions/3171540/long-constructor-initialization-lists

I guess new line for every parameter is fair enough.

DjWarmonger replied:
It's not that simple, as some constructors have dozens of parameters and take up a lot of vertical space.

dydzio0614 commented on AI/StupidAI/StupidAI.cpp:
```diff
@@ -117,7 +122,10 @@ BattleAction CStupidAI::activeStack( const CStack * stack )
-		static const std::vector<int> wallHexes = {50, 183, 182, 130, 78, 29, 12, 95};
+		static const std::vector<int> wallHexes =
+		{
+			50, 183, 182, 130, 78, 29, 12, 95
```
So here we have bit longer "list initializer"... When looking at it I do not find it to be particularly easier to read, and if the benefit is saving space then it will not help for really long value lists or when they are not short numbers.

dydzio0614 commented on AI/VCAI/VCAI.cpp:
```diff
@@ -3325,30 +3338,30 @@ this functions returns one target tile or invalid tile. We will use it to poll p
-	const Sector *src = &infoOnSectors[sourceSector],
-		*dest = &infoOnSectors[destinationSector];
+	const Sector * src = &infoOnSectors[sourceSector],
+	* dest = &infoOnSectors[destinationSector];
```
Ugh, is this "* dest" thing intended?

dydzio0614 replied:
Nevermind, it thought there is ; at end of 1st line

ArseniyShestakov replied:
Just in case this one is inteded because there is no difference for uncrustify config between:
```
const Sector * src = &infoOnSectors[sourceSector],
* dest = &infoOnSectors[destinationSector];
```
Or:
```
const Sector * src = &infoOnSectors[sourceSector];
const Sector * dest = &infoOnSectors[destinationSector];
```

dydzio0614 commented on CCallback.h:
```diff
@@ -34,61 +34,61 @@ class IBattleCallback
-	virtual int mergeStacks(const CArmedInstance *s1, const CArmedInstance *s2, SlotID p1, SlotID p2)=0;//joins first stack to the second (creatures must be same type)
-	virtual int mergeOrSwapStacks(const CArmedInstance *s1, const CArmedInstance *s2, SlotID p1, SlotID p2) =0; //first goes to the second
-	virtual int splitStack(const CArmedInstance *s1, const CArmedInstance *s2, SlotID p1, SlotID p2, int val)=0;//split creatures from the first stack
+	virtual void recruitHero(const CGObjectInstance * townOrTavern, const CGHeroInstance * hero) = 0;
```
Maybe it makes sense to debate if leaving pure virtual function "0's" without spacebars could be useful for marking this particular "0" usage and potentially faster code understanding. Not worth it though in my opinion.

ArseniyShestakov replied:
Again I don't think I can control this with options.
I'll make a list of thing I supposedly couldn't change via configuration in PR description.

dydzio0614 commented on Global.h:
```diff
@@ -279,442 +279,397 @@ void inline handleException()
-		return value >= min && value <= max;
-	}
+//checks if a is between b and c
+template<typename t1, typename t2, typename t3>	bool isbetween(const t1 & value, const t2 & min, const t3 & max)
```
One more example of template "one lining" - this is not helpful at all imo. Moreover we go for two-liners with constructor's initializer lists etc so why opposite approach.

dydzio0614 commented on Global.h:
```diff
@@ -279,442 +279,397 @@ void inline handleException()
+{
+	return std::unique_ptr<T>(new T(std::forward<Arg1>(arg1), std::forward<Arg2>(arg2), std::forward<Arg3>(arg3)));
+}
+template<typename T, typename Arg1, typename Arg2, typename Arg3, typename Arg4> std::unique_ptr<T> make_unique(Arg1 && arg1, Arg2 && arg2, Arg3 && arg3, Arg4 && arg4)
```
I definitely dont want to write code like this huge line

dydzio0614 commented on client/CBitmapHandler.cpp:
```diff
@@ -18,17 +18,17 @@
 }
 
-bool isPCX(const ui8 *header)//check whether file can be PCX according to header
+bool isPCX(const ui8 * header) //check whether file can be PCX according to header
```
Spacebar before comment - interesting change, i think code looks bit more tidy with it.

dydzio0614 commented on client/CBitmapHandler.cpp:
```diff
@@ -42,33 +42,36 @@ SDL_Surface * BitmapHandler::loadH3PCX(ui8 * pcx, size_t size)
 	Epcxformat format;
-	int it=0;
-
-	ui32 fSize = read_le_u32(pcx + it); it+=4;
```
What is current opinion on "never keep multiple statements in one line?" I wonder if these "multiple statements" dont actually improve code readability when used smartly.

IvanSavenko replied:
IMO this indicates different problem - `it+=4` should be part of `read_le_u32` - e.g. wrapped in some memory buffer class instead of manually advancing pointer for correct length each time you call this method.

dydzio0614 replied:
This is example of general case when sometimes multi statements may have their use, but yep...
I guess i am getting too paranoid about theoretical cases while our codebase has almost 0 legit usage of "multi statements in one line"

dydzio0614 commented on client/CBitmapHandler.cpp:
```diff
@@ -109,43 +112,43 @@ SDL_Surface * BitmapHandler::loadBitmapFromDir(std::string path, std::string fna
-	if (isPCX(readFile.first.get()))
-	{//H3-style PCX
+	if(isPCX(readFile.first.get()))
+	{ //H3-style PCX
```
Is it decided yet where are we allowed to place comments? For example Microsoft's C# guidelines discourage using comments at the end of lines with code, in favor of comments one line above code. I do not that too much though.

DjWarmonger replied:
Opening brace is a perfect place for a comment - it carries no other message, but lets us save some vertical space.

dydzio0614 commented on client/CBitmapHandler.cpp:
```diff
@@ -109,43 +112,43 @@ SDL_Surface * BitmapHandler::loadBitmapFromDir(std::string path, std::string fna
-		          1); // mark it for auto-deleting
-		if (ret)
+			//create SDL_RW with our data (will be deleted by SDL)
+			SDL_RWFromConstMem((void *)readFile.first.get(), readFile.second),
```
Not sure what to think about spacebar in casting, Is this side-effect work of uncrustify?

ArseniyShestakov replied:
With config we have now this is intended, but I not sure if this possible to adjust space before the pointer just for casting.

dydzio0614 commented on client/CMessage.h:
```diff
@@ -21,10 +21,10 @@ class CMessage
 public:
 	//Function usd only in CMessage.cpp
-	static std::pair<int,int> getMaxSizes(std::vector<std::vector<SDL_Surface*> > * txtg, int fontHeight);
+	static std::pair<int, int> getMaxSizes(std::vector<std::vector<SDL_Surface *>> * txtg, int fontHeight);
```
I think I reported this case already, but this change looks really dumb imo with that pointer-spacebar mangling.

dydzio0614 commented on client/battle/CBattleInterfaceClasses.h:
```diff
@@ -51,9 +52,10 @@ class CBattleConsole : public CIntObject
 public:
 	bool flip; //false if it's attacking hero, true otherwise
-	CDefHandler *dh, *flag; //animation and flag
+	CDefHandler * dh, * flag; //animation and flag
```
A bit unintuitive pointer init formatting

IvanSavenko replied:
Don't declare multiple members or variables in a single line and this won't be a problem.

dydzio0614 replied:
@ArseniyShestakov so maybe we should add some rules to coding guidelines about multi elements intantiation with comma?

ArseniyShestakov replied:
We'll certainly have them. We talked about that with @alexvins privately and there is certainly very few cases where you have legit reason to use comma in declaration.

I commented on it to Tow: https://github.com/vcmi/vcmi/pull/347#discussion_r127613322

dydzio0614 reviewed (commented):
My impressions on various changes after formatting...

dydzio0614 commented:
Now comment that actually answers SXX questions:

*Switching to no space after * and &.

-I think I will stay neutral here after all. That means, I think no space formatting makes sense as it is more common style and visually differentiates pointers from multiply operators in code (int c = (int)a * b; for example). I do not want to get in the way of more productive VCMI developers who like existing style more as I know I can adapt. This may fix some weird stuff from coverity like std::vector<Cstack *> - who needs that spacebar inside template parameters ❓ 

*Namespace indention. Some parts of our code look especially ugly when it's disabled.

-No personal opinion here, though what Tow said on slack sounds kinda good "at first glance" - no indenting for huge namespaces to make code within it feel naturally formatted. Unfortunately coverity is "all or nothing" tool.

ArseniyShestakov commented on AI/BattleAI/BattleAI.cpp:
```diff
@@ -53,7 +53,7 @@ BattleAction CBattleAI::activeStack( const CStack * stack )
 		{
 			auto healingTargets = cb->battleGetStacks(CBattleInfoEssentials::ONLY_MINE);
-			std::map<int, const CStack*> woundHpToStack;
+			std::map<int, const CStack *> woundHpToStack;
```
This is totally intended and I don't see anything wrong with that since ```*>``` on it's own have no special meaning in C++ as far as I aware. Space after the ptr star is questionable, but we'll have space before it for sure.

ArseniyShestakov commented on AI/BattleAI/common.h:
```diff
@@ -11,8 +11,7 @@
 
-template<typename Key, typename Val, typename Val2>
-const Val getValOr(const std::map<Key, Val> &Map, const Key &key, const Val2 defaultValue)
+template<typename Key, typename Val, typename Val2> const Val getValOr(const std::map<Key, Val> & Map, const Key & key, const Val2 defaultValue)
```
Depend on what other think we might not change it at all or force new line instead.
Problem is that half of our code use new line and other half is doesn't.

This is Uncrustify option:
```nl_template_class               = remove   # ignore/add/remove/force```

dydzio0614 replied:
I would say... always use new line. Will be easier to parse in head then as well.

ArseniyShestakov commented on AI/BattleAI/BattleAI.cpp:
```diff
@@ -286,7 +301,7 @@ void CBattleAI::attemptCastingSpell()
 	for(PossibleSpellcast & psc : possibleCasts)
 		psc.value = evaluateSpellcast(psc);
-	auto pscValue = [] (const PossibleSpellcast &ps) -> int
+	auto pscValue = [](const PossibleSpellcast & ps) -> int
```
I don't think Uncrustify provide any control over this, at least I didn't manage to find it.

dydzio0614 replied:
Just in case - resharper allows it if needed.

FeniksFire commented:
I can't find it anywhere so, what will be the format of `int **a`? This will not be something like `int * * a`?

ArseniyShestakov commented:
@FeniksFire yep it's a mess. Certainly the problem. Add it to roadblocks.

dydzio0614 commented:
I think target syntax should be int ** a; if we stay with current spacebar rules.
But anyway, using double pointers at this point will most likemy be some code smell... we do not write C and C++ has its own syntax for many different stuff.

ArseniyShestakov commented on server/CVCMIServer.cpp:
```diff
@@ -607,11 +608,11 @@ void handleLinuxSignal(int sig)
 #endif
 
-int main(int argc, char** argv)
+int main(int argc, char * * argv)
```
Let's comment on problem with ```***``` ```**``` here.

ArseniyShestakov replied:
@dydzio0614 for something like ```***``` we should probably use boost::multi_array, but we really only have 13 matches for ``` * * ``` now.

@FeniksFire do you think we should still care about them?

dydzio0614 replied:
To think that problem would be gone if we swapped to "no spacebar" style :D

FeniksFire replied:
Ehh, well sometimes it's nice to have double pointer, but if we have only 13 matches for them, then we can change it manually. In any case we should think about such things :).

ArseniyShestakov replied:
@FeniksFire yeah we certainly should keep it in mind and this is why this PR is here.
It's simply to hard to keep it all in mind and check it all for just one person.

Might be you can find some other weird bits of code here.

mwu-tow replied:
In this particular case — `argv` type should be actually `char *argv[]`.

dydzio0614 commented:
There is option that if we decide to unify some stuff uncrustify cant handle, we might use resharper and then decide what to do with these cases for future maintenance etc. Even if we dont find automaton solution it may be better to unify some stuff (like spacebars around lambda's ->) than leave "as is" because "uncrustify cannot handle"

ArseniyShestakov commented:
@dydzio0614 we can only use one solution for formatting because it's enough burden to maintain one. If Resharper will ever be useful on CI as command-line tool we can switch to it since I don't really care if utility is proprietary as long as it's get job done.

Uncrustify provide option to "ignore" some constructions that it doesn't handle well, but it's not always possible because some things like pure virtual functions for instance likely since they tied assignment in general.

dydzio0614 commented:
@SXX I mean to use resharper one time to unify some additional formatting stuff that we will not care about later because our standard tools will not be able check them. So we will end up with more consistent codebase for existing code. Lack of checks for these "stuff" for new code is another topic.

ArseniyShestakov commented on client/windows/CSpellWindow.cpp:
```diff
@@ -449,9 +444,13 @@ void CSpellWindow::keyPressed(const SDL_KeyboardEvent & key)
 			int index = -1;
-			while(schoolsOrder[++index] != selectedTab);
+			while(schoolsOrder[++index] != selectedTab)
+				;
```
This is something uncrustify doesn't handle well, but this probably need to be rewritten anyway.

mwu-tow replied:
Actually I like it that way — it is more explicit that there is an intended empty loop;

mwu-tow commented:
To answer the questions from the OP:
* I am supportive of "no space after * and &" change
* if we are to either have all namespaces indented or not indented, then I'd say "indent" (though it is depending on the fact that we have few namespaces, and hardly any nested ones)

alexvins commented on client/CPreGame.cpp:
```diff
@@ -181,77 +181,77 @@ void updateStartInfo(std::string filename, StartInfo & sInfo, const std::unique_
 };
 
-template <> class CApplyOnPG<CPack> : public CBaseForPGApply
+template<> class CApplyOnPG<CPack>: public CBaseForPGApply
```
Unwanted change. There should be spaces around `:` even after `>`

IvanSavenko commented on client/CVideoHandler.cpp:
```diff
@@ -185,48 +185,62 @@ bool CVideoPlayer::open(std::string fname, bool loop, bool useOverlay, bool scal
-							 SWS_BICUBIC, nullptr, nullptr, nullptr);
+		pos.w, pos.h,
+		AV_PIX_FMT_YUV420P,
+		SWS_BICUBIC, nullptr, nullptr, nullptr);
```
Why indention has been removed here? Now you don't see that these 3 lines are actually parameters from previous method.

ArseniyShestakov replied:
This is not really intended and there must be single tab indention. Probably I should really leave all lambdas completely broken for now because  "indent_continue = -1" confuse everyone.

alexvins commented on client/CPreGame.cpp:
```diff
@@ -1325,7 +1356,8 @@ SelectionTab::SelectionTab(CMenuScreen::EState Type, const std::function<void(CM
 	case CMenuScreen::loadGame:
-	case CMenuScreen::saveGame:;
+	case CMenuScreen::saveGame:
+		;
```
to be removed manually

dydzio0614 replied:
I personally like this change

ArseniyShestakov replied:
@dydzio0614 Alex mean that extra semicolon ";"

dydzio0614 replied:
ah, "extra". Okay

IvanSavenko commented on client/battle/CCreatureAnimation.cpp:
```diff
@@ -425,28 +427,22 @@ inline void CCreatureAnimation::putPixel(ui8 * dest, const SDL_Color & color, si
-	return getType() == CCreatureAnim::MOVE_START
-	    || getType() == CCreatureAnim::MOVING
-	    || getType() == CCreatureAnim::MOVE_END;
+	return getType() == CCreatureAnim::MOVE_START || getType() == CCreatureAnim::MOVING || getType() == CCreatureAnim::MOVE_END;
```
Why this was merged into a single line?
Not sure how common large ` A || B || C && D` statements in code but having each element on a separate line usually looks better.

ArseniyShestakov replied:
Basically for the reason mentioned in PR description: there is serious issue that Uncrustify try to indent multi-line conditions and align them with spaces. This is controlled following options:
```
# The position of boolean operators in wrapped expressions
pos_bool                        = join     # ignore/join/lead/lead_break/lead_force/trail/trail_break/trail_force
```

Other than that almost all VCMI code that using multi-line condition is just bad or totally horrible: especially CGameHandler with complain and comments inside if-else conditions. It's so bad so in some cases something like that is preferable:
```
bool boolOptionNameOne = somelongcodition == somelongtype;
bool boolOptionNameTwo = somelongcodition == somelongtype && 9000 == 9000;
if(boolOptionNameOne && boolOptionNameTwo)
```
To what we have how:
```
			//ranged counterattack
			if (destinationStack->hasBonusOfType(Bonus::RANGED_RETALIATION)
				&& !stack->hasBonusOfType(Bonus::BLOCKS_RANGED_RETALIATION)
				&& destinationStack->ableToRetaliate()
				&& stack->alive()) //attacker may have died (fire shield)
			{
```

dydzio0614 replied:
I think this particular code could be refactored to switch(getType())... for the sake of readability after reformatting

ArseniyShestakov commented on AI/StupidAI/StupidAI.cpp:
```diff
@@ -94,20 +96,23 @@ bool isCloser(const EnemyInfo & ei1, const EnemyInfo & ei2, const ReachabilityIn
+	}; //count of shooters on hexes
 
 	for(int i = 0; i < 2; i++)
-		for (auto & neighbour : (i ? h2 : h1).neighbouringTiles())
```
This is something need separate fixing.

ArseniyShestakov commented on AI/VCAI/AIUtility.cpp:
```diff
@@ -126,7 +128,7 @@ const CGHeroInstance * HeroPtr::operator*() const
 }
 
-void foreach_tile_pos(std::function<void(const int3& pos)> foo)
+void foreach_tile_pos(std::function<void(const int3 &pos)> foo)
```
Looks like a bug. It's all formatted automatically.

dydzio0614 replied:
maybe the digit "3" is source of bug. However with bugs like this... that source file will get bugged here everytime its processed by uncrustify

ArseniyShestakov commented on AI/VCAI/VCAI.cpp:
```diff
@@ -703,7 +703,8 @@ void VCAI::showGarrisonDialog(const CArmedInstance *up, const CGHeroInstance *do
 	status.addQuery(askID, "Map object select query");
-	requestActionASAP([=]{ answerQuery(askID, 0); });
+	requestActionASAP([ = ] { answerQuery(askID, 0);
+	});
```
Yes.

ArseniyShestakov commented on client/CDefHandler.cpp:
```diff
@@ -152,102 +153,104 @@ SDL_Surface * CDefHandler::getSprite (int SIndex, const ui8 * FDef, const SDL_Co
-	if (TopMargin > 0 || BottomMargin > 0 || LeftMargin > 0 || RightMargin > 0) {
-		memset( reinterpret_cast<char*>(ret->pixels), 0, FullHeight*FullWidth);
+	if(TopMargin > 0 || BottomMargin > 0 || LeftMargin > 0 || RightMargin > 0)
+	{
```
It did not add {}. All code modification options are disabled.

ArseniyShestakov commented on Global.h:
```diff
@@ -279,442 +279,397 @@ void inline handleException()
+template<typename t1, typename t2> t1 & amax(t1 & a, const t2 & b)
+{
+	if(a >= b)
+		return a;
```
Don't think there way to enforce something like that, but I'll add it to wiki.

dydzio0614 commented on AI/VCAI/Goals.cpp:
```diff
@@ -883,43 +879,43 @@ TSubgoal GatherTroops::whatToDoToAchieve()
 			BuildingID bid(BuildingID::DWELL_FIRST + creature->level - 1 + upgradeNumber * GameConstants::CREATURES_PER_TOWN);
-			if (t->hasBuilt(bid)) //this assumes only creatures with dwellings are assigned to faction
+			if(t->hasBuilt(bid)) //this assumes only creatures with dwellings are assigned to faction
 			{
```
this change only fixed spacebar - excessive { } is separate problem. This commit only changed whitespaces.

ArseniyShestakov commented on client/CMT.cpp:
```diff
@@ -1295,9 +1292,12 @@ static void handleEvent(SDL_Event & ev)
+	});
 
-	inGuiThread.reset(new bool(true));
+	inGuiThread.reset(new bool (true));
```
I think uncrustify would do the same for any of standard types. I suppose it's might be a bug.

ArseniyShestakov commented on Global.h:
```diff
@@ -279,442 +279,397 @@ void inline handleException()
-	}
+//returns true if unordered set c contains item i
+template<typename Item>	bool contains(const std::unordered_set<Item> & c, const Item & i)
+{
```
This is problem with all of "nl_" options and ```nl_template_class``` in particular. Basically when newline removed it's keep that extra tab untouched even in case of colons and such. Se we'll need to manually replace some tabs with spaces once.

Though we already decide we don't want template to be on same line everywhere.

ArseniyShestakov committed:
Uncrustify: add initial configuration

ArseniyShestakov committed:
Code style: uncrustify formatting preview

ArseniyShestakov commented:
I just pushed new preview with following changes:

```pos_bool = ignore```
- Joining of bool operators '&&' into one-line is disabled. This expose problem with indent of multi-line if-else conditions.
```
nl_cpp_lambda_leave_one_liners = true
nl_assign_leave_one_liners = true
```
- Formatting of single-line lambdas and assignments wouldn't be touched now, but we added [new rules regarding lambdas usage on wiki](https://wiki.vcmi.eu/index.php?title=Coding_guidelines#Lambda_expressions).
```nl_template_class = ignore```
- New line presence after template<> will be ignored. It's up to developer to add one or keep it same line.
```align_keep_tabs = false```
- Will remove some of useless spaces remaining after new line removal, e.g one going after comma in lists in place of space.
```indent_else_if = true```
- Will convert ambiguous if-else indent that don't have braces. E.g I added braces around two of such statements in cdd9dbc7f1228f62b872e800cca301e0ee38fb19. Also rule about that added to wiki that if indention is required then braces should be present.

**Lambdas are still broken and multi-line conditions have wrong indent!**

ArseniyShestakov commented on lib/mapping/CMapEditManager.h:
```diff
@@ -58,8 +58,9 @@ template<typename T>
-	explicit CMapSelection(CMap * map) : map(map) { }
-	virtual ~CMapSelection() { };
+	explicit CMapSelection(CMap * map)
+		: map(map) {}
```
Some imperfection here, but might be we need no constructors with initializers lists in headers?

alexvins replied:
Impossible for templates, problematic for netpacks

alexvins commented on AI/VCAI/Fuzzy.cpp:
```diff
@@ -145,14 +145,14 @@ void FuzzyHelper::initTacticalAdvantage()
-			fl::Trapezoid* medium = new fl::Trapezoid("MEDIUM", (CGTownInstance::FORT - CGTownInstance::NONE) * 0.5f, CGTownInstance::FORT,
-				CGTownInstance::CITADEL, CGTownInstance::CITADEL + (CGTownInstance::CASTLE - CGTownInstance::CITADEL) * 0.5f);
+			fl::Trapezoid * medium = new fl::Trapezoid("MEDIUM", (CGTownInstance::FORT - CGTownInstance::NONE) * 0.5f, CGTownInstance::FORT,
+								   CGTownInstance::CITADEL, CGTownInstance::CITADEL + (CGTownInstance::CASTLE - CGTownInstance::CITADEL) * 0.5f);
```
I don`t like such aligning, even if it would be accepted align should be done with spaces

ArseniyShestakov replied:
I don't like any alignment at all especially with spaces, but I still not find a way to control it in every case.
